sw/device/lib/dif/README.md

```diff
@@ -81,6 +81,31 @@ Notational caveats:
 * Unless otherwise noted, all symbols mentioned below are required.
 
 
+#### Hardware Parameterization
+
+Our aim is that a single DIF library can be used with multiple instances of the
+same IP on the same chip, even when those IPs have been instantiated with
+different hardware parameters.
+
+At the moment, we have a good approach to being able to address separate
+hardware instances instantiated at separate addresses, as long as they have the
+same hardware parameters (see the `base_addr` member in `dif_<ip>_params_t`).
+Most other parameters come from the specific IP on a case-by-case basis.
+
+As much as possible, we would like to hide these parameters underneath the
+interface of the DIF, but this is not always possible, especially where
+particular functionality requires the DIF caller to allocate memory.
+
+In order to support exposing these parameters to callers, DIFs should provide
+query functions which take a `dif_<ip>_params_t`, rather than `#define`s or
+global variable definitions. This should be done even if the DIF does not
```

Comment:
I wonder if it would be better just to use the params struct passed to init (which is stored in the device instance struct) for this purpose instead of handling this in a per-query-function basis.

Reply:
I guess it might be nicer for the caller with the explicit API, and gives you some more flexibility in case the implementation has to modify the values. Presumably these would be non fail simple getters, so not too much extra work.

Reply:
I don't want this to take a `dif_<ip>_t`, because of peripherals like USBDev - we end up needing to allocate memory to create one of those handles (at the moment we use constants, this PR is about the guidance so we can move away from that) - and you need to allocate that buffer before you call `init`, so you can pass it to `init`, which is what returns a `dif_<ip>_t` (as i envision that API changing to cope with parameterisation).

I think it's important to maintain that flexibility in these query functions. As @silvestrst these will hopefully just be getters, not functions that return a `result_t`.